src/common/prover/duties/bad-performers.service.ts

```diff
@@ -0,0 +1,317 @@
+    if (csmVersion == 1) return [];
+    if (!this.isV2Initialized) await this.initV2();
+    //
+    const strikesTree = await this.getStrikesTree(headBlockInfo);
+    if (!strikesTree) return [];
+    const thresholds = await this.getStrikesThresholds(headBlockInfo);
+    if (this.isStrikesTreeAlreadyProcessed(strikesTree.root) && !this.isAnyStrikesThresholdChanged(thresholds)) {
+      return [];
+    }
+    this.currentStrikesTree = strikesTree;
+    this.currentStrikesThresholdsByCurveId = thresholds;
+    const badPerfKeys = await this.getBadPerformersKeys(headBlockInfo, fullKeyInfoFn);
+    if (!badPerfKeys) return [];
+    const unproven = await this.getUnprovenKeys(headBlockInfo, badPerfKeys);
+    if (!unproven) return [];
+    const unprovenNonWithdrawn = await this.getNonWithdrawnKeys(headBlockInfo, unproven);
+    if (!unprovenNonWithdrawn) return [];
+    return unprovenNonWithdrawn;
+  }
+
+  public async sendBadPerformanceProofs(badPerformers: InvolvedKeysWithBadPerformance): Promise<number> {
+    if (!Object.keys(badPerformers).length) return 0;
+
+    if (!this.currentStrikesTree) {
+      throw new Error('Strikes Tree should be initialized before sending bad performance proofs');
+    }
+
+    const keysMaxBatchSize = this.config.get('TX_STRIKES_PAYLOAD_MAX_BATCH_SIZE');
+
+    const batchCount = Math.ceil(badPerformers.length / keysMaxBatchSize);
+
+    this.logger.log(
+      `Preparing payloads for ${badPerformers.length} validators in ${batchCount} batches by ${keysMaxBatchSize} max keys each`,
+    );
+
+    badPerformers.sort((a, b) => b.leafIndex - a.leafIndex);
+
+    await this.processBadPerformerBatches(badPerformers, keysMaxBatchSize);
+
+    this.lastProcessedStrikesTreeRoot = this.currentStrikesTree.root;
+    return badPerformers.length;
+  }
+
+  private async processBadPerformerBatches(
+    badPerformers: InvolvedKeysWithBadPerformance,
+    keysMaxBatchSize: number,
+  ): Promise<void> {
+    for (let i = 0; i < badPerformers.length; i += keysMaxBatchSize) {
+      const batch = badPerformers.slice(i, i + keysMaxBatchSize);
+      await this.processBatch(batch);
+    }
+  }
+
+  private async processBatch(batch: InvolvedKeysWithBadPerformance): Promise<void> {
+    if (!this.currentStrikesTree) {
+      throw new Error('Strikes Tree should be initialized before processing batches');
+    }
+
+    const leavesIndices = batch.map((key) => key.leafIndex);
+    const multiProof = this.currentStrikesTree.getMultiProof(leavesIndices);
+    const keyStrikesList = this.buildKeyStrikesPayload(multiProof.leaves, batch);
+
+    const validatorIndices = batch.map((key) => key.validatorIndex).join(', ');
+    this.logger.log(`📡 Sending bad performer multi-proof payload for batch of validators: ${validatorIndices}`);
+
+    await this.strikes.sendBadPerformanceProof({
+      keyStrikesList,
+      proof: multiProof.proof,
+      proofFlags: multiProof.proofFlags,
+    });
+  }
```

Comment:
I would merge the `processBadPerformerBatches` with `processBatch` and moved `badPerformers.sort` to `processBadPerformerBatches`, because I guess that's for better use of multiproofs.